### ISSUE TITLE
[feat] Snap layouts support on Windows when using frameless windows

### ISSUE THREAD

**Issue by grislux55** [tauri-apps/tauri#4531]
### Describe the problem

[Details about snap layouts](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/apply-snap-layout-menu)

If I use the window decorations provided by Tauri, then I can use snap layouts.

![snap layouts](https://user-images.githubusercontent.com/33774862/176607202-b11f4a1f-dcfc-49ea-a1be-788a4bfd8705.png)

If I set the window to `"decorations": false`, the window will not be able to use snap layouts.

![no snap layouts](https://user-images.githubusercontent.com/33774862/176606655-34d18560-7c4e-40c4-b70b-9fa628b60ccf.png)

### Describe the solution you'd like

Tauri already uses data-tauri-drag-region to support window dragging, so we might consider introducing a attribute like data-tauri-maximize-button-region to support this feature.

This should work as follows.

```
<div data-tauri-drag-region class="titlebar">
  --- omitted ---
  <div data-tauri-maximize-button-region class="titlebar-button" id="titlebar-maximize">
    <img
      src="https://api.iconify.design/mdi:window-maximize.svg"
      alt="maximize"
    />
  </div>
  --- omitted ---
</div>
```

### Alternatives considered

Platform-specific window behavior is very complex, and leaving window behavior to the user can be tricky to implement.

Electron provides [BrowserWindow ](https://www.electronjs.org/docs/latest/api/browser-window) to handle similar issues.

Tauri can provide similar built-in components, which will also be easy to manage and optimize.

**Comment by Voidoz** [comment]
Would love to see this.
I noticed that in tao the WM_NCHITTEST message is ignored... Link below:
https://github.com/tauri-apps/tao/blob/dev/src/platform_impl/windows/window.rs#L1176

Could we use hover events over an element to decide what you return to that message? Could possibly even be used for all other return codes?

Forgive me if I'm completely off base. I'm very new to low level windows APIs and also am not at all familiar with the tao codebase.

**Comment by AZProductions** [comment]
Are there any plans on supporting snap layouts?

**Comment by FabianLars** [comment]
Yes, or maybe "wishes" instead of "plans". Anyway, the lack of information is my bad, i only ever talked about this in other issues and on discord. So here's a tldr:
The snap layout can only be openend by reacting to the NC_HITTEST message. Unfortunately these messages are not sent for clicks inside the webview. The webview2 team is working on an api proposal that should help us here though: https://github.com/MicrosoftEdge/WebView2Feedback/pull/3367 but who knows when it will be merged and actually implemented.

**Comment by mchao123** [comment]
Waiting for webview2 to open the api I think is irrational, although there are compatibility issues with adaptation.

**Comment by hazre** [comment]
> Yes, or maybe "wishes" instead of "plans". Anyway, the lack of information is my bad, i only ever talked about this in other issues and on discord. So here's a tldr: The snap layout can only be openend by reacting to the NC_HITTEST message. Unfortunately these messages are not sent for clicks inside the webview. The webview2 team is working on an api proposal that should help us here though: [MicrosoftEdge/WebView2Feedback#3367](https://github.com/MicrosoftEdge/WebView2Feedback/pull/3367) but who knows when it will be merged and actually implemented.

Looks like it has been merged now.

**Comment by AClon314** [comment]
Wish it will be implemented in the standard tauri v2 :)

**Comment by FabianLars** [comment]
This won't be implemented for v2 since we're still blocked by webview2. The pr from their repo I linked above was sadly for their composition based apis only and we currently can't use those.

**Comment by Daydreamer-riri** [comment]
‌‌Can we raise our issues to them to gain support? Is there any such action currently?

**Comment by FabianLars** [comment]
> ‌‌Can we raise our issues to them to gain support? Is there any such action currently?

There were a few issues already, especially <https://github.com/MicrosoftEdge/WebView2Feedback/issues/446>. There was no interest in fixing this (or adding support for it, however you wanna phrase it) in a more general way. They are working on WCO though: https://github.com/MicrosoftEdge/WebView2Feedback/issues/4532 but that won't support win 7/8 if that matters.

**Comment by stevietelly** [comment]
anything new now?

**Comment by mchao123** [comment]
I try to listen for mouse events and simulate keyboard WIN+Z to achieve a close layout, but there seems to be a problem with the location. Does anyone know what to do?
<img width="350" alt="image" src="https://github.com/MicrosoftEdge/WebView2Feedback/assets/24653378/9286a544-0df1-4e5b-9981-0370cde48bab">

**Comment by grislux55** [comment]
> I try to listen for mouse events and simulate keyboard WIN+Z to achieve a close layout, but there seems to be a problem with the location. Does anyone know what to do?

@mchao123 I guess because Tauri uses a special window structure to implement drag capability for frameless windows ([attach_resize_handler](https://github.com/tauri-apps/tauri/blob/2d029a9f53117f756f9d6783fab74b06a7ab3e8e/crates/tauri-runtime-wry/src/undecorated_resizing.rs#L117), [set_drag_hwnd_rgn](https://github.com/tauri-apps/tauri/blob/2d029a9f53117f756f9d6783fab74b06a7ab3e8e/crates/tauri-runtime-wry/src/undecorated_resizing.rs#L413)), Windows cannot correctly determine the actual position of the window.

**Comment by SFINXVC** [comment]
I have it working in my own fork of Tauri, using the same approach Tauri uses to handle the drag region.
<img width="1140" height="828" alt="Image" src="https://github.com/user-attachments/assets/955dd042-3001-47c0-b7d6-e9d2e0df94be" />

**Comment by PleaseDont** [comment]
@SFINXVC Where's the code?

**Comment by dunkyl** [comment]
I was able to do a workaround, probably similar to what SFINXVC has done, but without a Tauri fork. [Here is the code](https://github.com/dunkyl/tauri-snap-layout) (in Dioxus but just as applicable to other Tauri apps).

**Comment by Zbrooklyn** [comment]
Adding to what @dunkyl and @SFINXVC posted, since this thread is where most people land: the child-window approach works today with **no Tauri fork**, and I've put up two runnable templates plus a full write-up of why every other approach fails.

**https://github.com/Zbrooklyn/tauri-snap-layouts**

Two things I'd flag for anyone arriving here, because they cost me the most time and I haven't seen them written down anywhere:

**1. `status: upstream` on this issue does not mean your app is blocked.**

It blocks Tauri from shipping a *first-class API*. It says nothing about what your app can do. Nothing stops you creating your own Win32 child window alongside Tauri's, and that is what every working solution in this thread does. I think this misreading is why the thread is four years old with people still concluding "use Electron."

**2. Subclassing the Tauri window to answer `WM_NCHITTEST` is structurally impossible, not merely difficult.**

This is worth stating plainly because it is *literally the code in Microsoft's documentation*, so people reasonably try it first and lose days. Your page lives in a WebView2 child window (`Chrome_RenderWidgetHostHWND`) covering the whole client area, so it answers the hit-test before your window ever sees it. Chromium also uses its own input routing internally, so `SetWindowSubclass` on the descendants that actually receive the mouse fails outright.

You can recognise you're in this trap because your maximize button **still shows its CSS `:hover` and its HTML `title` tooltip** — both of those require the webview to have received the mouse, which means your window did not.

The working shape is a small transparent native child window sitting exactly over your HTML maximize button, whose window proc returns `HTMAXBUTTON` unconditionally. It never paints, so your design shows through, but it owns the mouse there — so it has to report hover and clicks back to your page.

```rust
tauri::Builder::default()
    .plugin(
        tauri_plugin_frame::FramePluginBuilder::new()
            .auto_titlebar(false)
            .snap_overlay(true)
            .titlebar_height(32)   // MUST match your CSS
            .button_width(46)      // MUST match your CSS
            .build(),
    )
    .setup(|app| {
        let window = app.get_webview_window("main").unwrap();
        window.create_overlay_titlebar_with_height(32)?;
        Ok(())
    })
```

```jsonc
// tauri.conf.json - minWidth is a Microsoft requirement, not a style choice
{ "decorations": false, "minWidth": 330 }
```

That `minWidth` line is the one that gets people after everything else is right: Microsoft asks for ≤500 effective pixels (ideally ≤330), and above it **the flyout appears but the window won't actually snap into a zone**. It looks like a half-broken feature rather than a config problem.

The repo has two templates (a plain one-canvas shell, and a Windows Terminal-style tab strip sharing the row with the caption buttons), rounded corners via `DWMWA_WINDOW_CORNER_PREFERENCE`, and a `verify.ps1` that asserts the overlay is really hit-testing the maximize button on a live window rather than trusting a screenshot — `PrintWindow` cannot capture the flyout at all, since the flyout is a separate OS window, so screenshot-based verification produces confident false failures.

Verified on Windows 11 build 26200 / Tauri 2.11.5, at both 100% and 150% scaling.
